packages/block-library/src/navigation/index.php

```diff
@@ -632,11 +647,17 @@ function render_block_core_navigation( $attributes, $content, $block ) {
 	$toggle_aria_label_close     = $should_display_icon_label ? 'aria-label="' . __( 'Close menu' ) . '"' : ''; // Close button label.
 
 	$responsive_container_markup = sprintf(
-		'<button aria-haspopup="true" %3$s class="%6$s" data-micromodal-trigger="%1$s">%9$s</button>
-			<div class="%5$s" style="%7$s" id="%1$s">
-				<div class="wp-block-navigation__responsive-close" tabindex="-1" data-micromodal-close>
-					<div class="wp-block-navigation__responsive-dialog" aria-label="%8$s">
-							<button %4$s data-micromodal-close class="wp-block-navigation__responsive-container-close">%10$s</button>
+		'<button x-on:click="open = true; $focus.within($refs.items).first()" aria-haspopup="true" %3$s class="%6$s">%9$s</button>
```

Comment:
A few things here with the `$focus.within($refs.items)`:

- The `$focus.within($refs.items).first()` part is not triggered when you focus on the hamburguer menu and use `Enter` to open the menu.

  I wonder what directive/pattern we could use here to take into account both the click and the keyboard. Maybe a simple `x-effect` that reacts to `open === true` to make it more declarative?

- `$focus.within($refs.items).first()` is focusing the last element to me! I have no idea why.

  https://user-images.githubusercontent.com/3305402/191493205-e5b9ecf9-ebce-4925-8446-425a5e40f739.mp4

- On close, the `micromodal` version focus the hamburguer menu (shown in video above). We should include that.

- What do you think about the `$focus.within()` API. Was it easy to understand? Would it be an easier way to deal with the focus?

Reply:
>* The $focus.within($refs.items).first() part is not triggered when you focus on the hamburguer menu and use Enter to open the menu.
>
>I wonder what directive/pattern we could use here to take into account both the click and the keyboard. Maybe a simple x-effect that reacts to open === true to make it more declarative?
>
>* $focus.within($refs.items).first() is focusing the last element to me! I have no idea why.

I'm not sure about these two points. It seems to be working for me. Or am I doing something wrong?

https://www.loom.com/share/a144d46e28e6403ca77e2fa7674c87f2

> On close, the micromodal version focus the hamburguer menu (shown in video above). We should include that.

Regarding this, I believed it was working 🤷 I have just pushed a commit fixing it using `x-ref` and `$focus` again. Not sure if it is the best way though, maybe we could use just something like `document.querySelector(".hamburger").focus()`.

> What do you think about the $focus.within() API. Was it easy to understand? Would it be an easier way to deal with the focus?

For me, it was easy to understand but maybe it was just because there is an example really similar to this use case in the docs. If that wasn't the case, I would have probably struggled to understand it.

Regarding if there is a better way to handle the focus, maybe for these cases using the directives is not 100% necessary and we could use just `document.querySelector(".hamburger").focus()` as mentioned above.

Anyway, the [functions of the `$focus`](https://alpinejs.dev/plugins/focus#focus-magic) seem easy to use to me and cover many use cases.

Reply:
> I'm not sure about these two points. It seems to be working for me. Or am I doing something wrong?

Oh, you're right. Alpine uses `tappable`, and the [list of focusable elements](https://github.com/focus-trap/tabbable/blob/master/src/index.js#L1-L13) is not the same [as the one in micromodal](https://github.com/ghosh/Micromodal/blob/master/lib/src/index.js#L4-L16), so maybe that's why. Don't worry about it. It's not important.

> I wonder what directive/pattern we could use here to take into account both the click and the keyboard. Maybe a simple x-effect that reacts to open === true to make it more declarative?

I've moved the focus to an `x-effect` directive (which is similar to React's `useEffect`) in https://github.com/WordPress/gutenberg/pull/44289/commits/0a0598f147f3b413e6b223eb5acdb07bed153b29 to make it more declarative. It seems to work fine, but could you please check it out?

`x-effect="open === true ? $focus.within($refs.items).first() : $focus.focus($refs.hamburger)`

_I couldn't use `if ... else` inside the `x-effect`. I'm not sure if I was doing something wrong._

> Not sure if it is the best way though, maybe we could use just something like document.querySelector(".hamburger").focus().

Good point!

Reply:
> I've moved the focus to an x-effect directive (which is similar to React's useEffect) in https://github.com/WordPress/gutenberg/commit/0a0598f147f3b413e6b223eb5acdb07bed153b29 to make it more declarative. It seems to work fine, but could you please check it out?

Oh, I missed the `x-effect` comment sorry. It seems clearer this way. However, if I am not mistaken when the page is loaded `open = false` so it leads to focusing on the hamburger even if the user hasn't interacted with the menu. I assume this shouldn't be the case right? The only thing I can think of to solve it is to change the initial state of open to undefined or an empty string and change the conditional to only focus on true /false:

```
x-effect="if (open === true) { $focus.within($refs.items).first() } else if (open === false) { $focus.focus($refs.hamburger) }"
```

This way it seems to work for me. 

_And using the `if ... else` is working for me 🤷_

Reply:
> However, if I am not mistaken when the page is loaded open = false so it leads to focusing on the hamburger even if the user hasn't interacted with the menu

Oh, true. That's a bummer.

Reply:
> $focus.within($refs.items).first() is focusing the last element to me! I have no idea why.

I think I've discovered why. One of the items has `tabindex="0"` in the HTML. I don't know where did it come from.